Framework/Geometry/inc/MantidGeometry/Objects/MeshObject.h

```diff
@@ -0,0 +1,212 @@
+  Kernel::V3D generatePointInObject(Kernel::PseudoRandomNumberGenerator &rng,
+                                    const BoundingBox &activeRegion,
+                                    const size_t) const override;
+
+  // Rendering member functions
+  void draw() const override;
+  // Initialize Drawing
+  void initDraw() const override;
+  // Get Geometry Handler
+  boost::shared_ptr<GeometryHandler> getGeometryHandler() override;
+  /// Set Geometry Handler
+  void setGeometryHandler(boost::shared_ptr<GeometryHandler> h);
+
+  /// set vtkGeometryCache writer
+  void setVtkGeometryCacheWriter(
+      boost::shared_ptr<vtkGeometryCacheWriter>) override;
+  /// set vtkGeometryCache reader
+  void setVtkGeometryCacheReader(
+      boost::shared_ptr<vtkGeometryCacheReader>) override;
+  void GetObjectGeom(int &type, std::vector<Kernel::V3D> &vectors,
+                     double &myradius, double &myheight) const override;
+  /// Getter for the shape xml
+  std::string getShapeXML() const override;
+
+  /// Read access to mesh object for rendering
+  int numberOfVertices() const;
+  double *getVertices() const;
+  int numberOfTriangles() const;
+  int *getTriangles() const;
+
+  void updateGeometryHandler();
+
+private:
+  /// Default constructor - should never be called
+  MeshObject();
+  /// Get intersections
+  void getIntersections(const Kernel::V3D &start, const Kernel::V3D &direction,
+                        std::vector<Kernel::V3D> &intersectionPoints,
+                        std::vector<int> &entryExitFlags) const;
+  /// Determine intersection between ray and an one triangle
+  bool rayIntersectsTriangle(const Kernel::V3D &start,
+                             const Kernel::V3D &direction,
+                             const Kernel::V3D &v1, const Kernel::V3D &v2,
+                             const Kernel::V3D &v3, Kernel::V3D &intersection,
+                             int &entryExit) const;
+  /// Get triangle
+  bool getTriangle(const size_t index, Kernel::V3D &v1, Kernel::V3D &v2,
+                   Kernel::V3D &v3) const;
+  /// Search object for valid point
+  bool searchForObject(Kernel::V3D &point) const;
+
+  /// Cache for object's bounding box
+  mutable BoundingBox m_boundingBox;
+
+  /// Tolerence distance
+  const double M_TOLERANCE = 0.000001;
+
+  /// Numerical identifier of object
+  int m_object_number;
+
+  /// Geometry Handle for rendering
+  boost::shared_ptr<GeometryHandler> m_handler;
+
+  // String from which object may be defined
+  std::string m_string;
+
+  /// Optional string identifier
+  std::string m_id;
+  /// material composition
+  std::unique_ptr<Kernel::Material> m_material;
+
+  /// a pointer to a class for reading from the geometry cache
+  boost::shared_ptr<vtkGeometryCacheReader> m_vtk_cache_reader;
+  /// a pointer to a class for writing to the geometry cache
+  boost::shared_ptr<vtkGeometryCacheWriter> m_vtk_cache_writer;
+
+  /// Contents
+  std::vector<size_t> m_triangles;
```

Comment:
I wonder if defining this as `std::vector<uint16_t>` would be better. Since I cannot see for a MeshObject (defines a single shape), that 2^64 vertices would ever be used.  @martyngigg thoughts?

Reply:
I agree that we should use a fixed-size type. Is 65535 enough for say a laser scan of some complex sample? Would `uint32_t` be safer?

Reply:
Agreed. `uint32_t` would be safer. I don't see that we would use more than 65K points, but 4G points for a single shape should be easily enough to describe a single object. 

@KarlPalmen could you add a check in the constructor of `MeshObject` as part of these changes along the lines of:

```cpp
if (m_vertices.size() > std::numeric_limits<m_triangles::value_type>::max()) {
  throw std::invalid_argument("More vertices given than possible to triangulate"); // or some better message.
}
```
And add a test for this to `MeshObjectTest`.